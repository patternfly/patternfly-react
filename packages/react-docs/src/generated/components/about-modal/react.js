import React from 'react';
import { AutoLinkHeader, Example, Link as PatternflyThemeLink } from 'theme-patternfly-org/components';
import brandImg from '../../../../../react-core/src/components/AboutModal/examples/./brandImg.svg';
import bgImg from '../../../../../react-core/src/components/AboutModal/examples/./patternfly-orb.svg';
const pageData = {
  "id": "About modal",
  "section": "components",
  "source": "react",
  "slug": "/components/about-modal/react",
  "sourceLink": "https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/AboutModal/examples/AboutModal.md",
  "propComponents": [
    {
      "name": "AboutModal",
      "description": "",
      "props": [
        {
          "name": "appendTo",
          "type": "HTMLElement | (() => HTMLElement)",
          "description": "The parent container to append the modal to. Defaults to document.body",
          "defaultValue": "null"
        },
        {
          "name": "backgroundImageSrc",
          "type": "string",
          "description": "The URL of the image for the background",
          "defaultValue": "''"
        },
        {
          "name": "brandImageAlt",
          "type": "string",
          "description": "The alternate text of the brand image",
          "required": true
        },
        {
          "name": "brandImageSrc",
          "type": "string",
          "description": "The URL of the image for the brand",
          "required": true
        },
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Content rendered inside the about modal",
          "required": true
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the about modal",
          "defaultValue": "''"
        },
        {
          "name": "closeButtonAriaLabel",
          "type": "string",
          "description": "Set aria label to the close button"
        },
        {
          "name": "disableFocusTrap",
          "type": "boolean",
          "description": "Flag to disable focus trap"
        },
        {
          "name": "isOpen",
          "type": "boolean",
          "description": "Flag to show the about modal",
          "defaultValue": "false"
        },
        {
          "name": "noAboutModalBoxContentContainer",
          "type": "boolean",
          "description": "Prevents the about modal from rendering content inside a container; allows for more flexible layouts",
          "defaultValue": "false"
        },
        {
          "name": "onClose",
          "type": "() => void",
          "description": "A callback for when the close button is clicked",
          "defaultValue": "(): any => undefined"
        },
        {
          "name": "productName",
          "type": "string",
          "description": "Product name",
          "defaultValue": "''"
        },
        {
          "name": "trademark",
          "type": "string",
          "description": "Trademark information",
          "defaultValue": "''"
        }
      ]
    }
  ],
  "cssPrefix": [
    "pf-c-about-modal-box"
  ],
  "examples": [
    "Basic",
    "Without product name",
    "Complex user positioned content",
    "Custom background image"
  ]
};
pageData.liveContext = {
  brandImg,
  bgImg
};
pageData.examples = {
  'Basic': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { AboutModal, Button, TextContent, TextList, TextListItem } from '@patternfly/react-core';\nimport brandImg from './brandImg.svg';\n\nexport const AboutModalBasic: React.FunctionComponent = () => {\n  const [isModalOpen, setIsModalOpen] = React.useState(false);\n\n  const toggleModal = () => {\n    setIsModalOpen(!isModalOpen);\n  };\n\n  return (\n    <React.Fragment>\n      <Button variant=\"primary\" onClick={toggleModal}>\n        Show about modal\n      </Button>\n      <AboutModal\n        isOpen={isModalOpen}\n        onClose={toggleModal}\n        trademark=\"Trademark and copyright information here\"\n        brandImageSrc={brandImg}\n        brandImageAlt=\"Patternfly Logo\"\n        productName=\"Product Name\"\n      >\n        <TextContent>\n          <TextList component=\"dl\">\n            <TextListItem component=\"dt\">CFME Version</TextListItem>\n            <TextListItem component=\"dd\">5.5.3.4.20102789036450</TextListItem>\n            <TextListItem component=\"dt\">Cloudforms Version</TextListItem>\n            <TextListItem component=\"dd\">4.1</TextListItem>\n            <TextListItem component=\"dt\">Server Name</TextListItem>\n            <TextListItem component=\"dd\">40DemoMaster</TextListItem>\n            <TextListItem component=\"dt\">User Name</TextListItem>\n            <TextListItem component=\"dd\">Administrator</TextListItem>\n            <TextListItem component=\"dt\">User Role</TextListItem>\n            <TextListItem component=\"dd\">EvmRole-super_administrator</TextListItem>\n            <TextListItem component=\"dt\">Browser Version</TextListItem>\n            <TextListItem component=\"dd\">601.2</TextListItem>\n            <TextListItem component=\"dt\">Browser OS</TextListItem>\n            <TextListItem component=\"dd\">Mac</TextListItem>\n          </TextList>\n        </TextContent>\n      </AboutModal>\n    </React.Fragment>\n  );\n};\n","title":"Basic","lang":"ts"}}>
      
    </Example>,
  'Without product name': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { AboutModal, Button, TextContent, TextList, TextListItem } from '@patternfly/react-core';\nimport brandImg from './brandImg.svg';\n\nexport const AboutModalWithoutProductName: React.FunctionComponent = () => {\n  const [isModalOpen, setIsModalOpen] = React.useState(false);\n\n  const toggleModal = () => {\n    setIsModalOpen(!isModalOpen);\n  };\n\n  return (\n    <React.Fragment>\n      <Button variant=\"primary\" onClick={toggleModal}>\n        Show about modal\n      </Button>\n      <AboutModal\n        isOpen={isModalOpen}\n        onClose={toggleModal}\n        trademark=\"Trademark and copyright information here\"\n        brandImageSrc={brandImg}\n        brandImageAlt=\"Patternfly Logo\"\n      >\n        <TextContent>\n          <TextList component=\"dl\">\n            <TextListItem component=\"dt\">CFME Version</TextListItem>\n            <TextListItem component=\"dd\">5.5.3.4.20102789036450</TextListItem>\n            <TextListItem component=\"dt\">Cloudforms Version</TextListItem>\n            <TextListItem component=\"dd\">4.1</TextListItem>\n            <TextListItem component=\"dt\">Server Name</TextListItem>\n            <TextListItem component=\"dd\">40DemoMaster</TextListItem>\n            <TextListItem component=\"dt\">User Name</TextListItem>\n            <TextListItem component=\"dd\">Administrator</TextListItem>\n            <TextListItem component=\"dt\">User Role</TextListItem>\n            <TextListItem component=\"dd\">EvmRole-super_administrator</TextListItem>\n            <TextListItem component=\"dt\">Browser Version</TextListItem>\n            <TextListItem component=\"dd\">601.2</TextListItem>\n            <TextListItem component=\"dt\">Browser OS</TextListItem>\n            <TextListItem component=\"dd\">Mac</TextListItem>\n          </TextList>\n        </TextContent>\n      </AboutModal>\n    </React.Fragment>\n  );\n};\n","title":"Without product name","lang":"ts"}}>
      
    </Example>,
  'Complex user positioned content': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { AboutModal, Alert, Button, TextContent, TextList, TextListItem } from '@patternfly/react-core';\nimport brandImg from './brandImg.svg';\n\nexport const AboutModalComplexUserPositionedContent: React.FunctionComponent = () => {\n  const [isModalOpen, setIsModalOpen] = React.useState(false);\n\n  const toggleModal = () => {\n    setIsModalOpen(!isModalOpen);\n  };\n\n  return (\n    <React.Fragment>\n      <Button variant=\"primary\" onClick={toggleModal}>\n        Show about modal\n      </Button>\n      <AboutModal\n        isOpen={isModalOpen}\n        onClose={toggleModal}\n        trademark=\"Trademark and copyright information here\"\n        brandImageSrc={brandImg}\n        brandImageAlt=\"Patternfly Logo\"\n        noAboutModalBoxContentContainer={true}\n        productName=\"Product Name\"\n      >\n        <TextContent id=\"test1\" className=\"pf-u-py-xl\">\n          <h4>About</h4>\n          <p>Content here</p>\n        </TextContent>\n        <Alert variant=\"info\" title=\"Updates available\" />\n        <TextContent id=\"test2\" className=\"pf-u-py-xl\">\n          <TextList component=\"dl\">\n            <TextListItem component=\"dt\">CFME Version</TextListItem>\n            <TextListItem component=\"dd\">5.5.3.4.20102789036450</TextListItem>\n            <TextListItem component=\"dt\">Cloudforms Version</TextListItem>\n            <TextListItem component=\"dd\">4.1</TextListItem>\n            <TextListItem component=\"dt\">Server Name</TextListItem>\n            <TextListItem component=\"dd\">40DemoMaster</TextListItem>\n            <TextListItem component=\"dt\">User Name</TextListItem>\n            <TextListItem component=\"dd\">Administrator</TextListItem>\n            <TextListItem component=\"dt\">User Role</TextListItem>\n            <TextListItem component=\"dd\">EvmRole-super_administrator</TextListItem>\n            <TextListItem component=\"dt\">Browser Version</TextListItem>\n            <TextListItem component=\"dd\">601.2</TextListItem>\n            <TextListItem component=\"dt\">Browser OS</TextListItem>\n            <TextListItem component=\"dd\">Mac</TextListItem>\n          </TextList>\n        </TextContent>\n      </AboutModal>\n    </React.Fragment>\n  );\n};\n","title":"Complex user positioned content","lang":"ts"}}>
      
    </Example>,
  'Custom background image': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { AboutModal, Button, TextContent, TextList, TextListItem } from '@patternfly/react-core';\nimport brandImg from './brandImg.svg';\nimport bgImg from './patternfly-orb.svg';\n\nexport const AboutModalCustomBackgroundImage: React.FunctionComponent = () => {\n  const [isModalOpen, setIsModalOpen] = React.useState(false);\n\n  const toggleModal = () => {\n    setIsModalOpen(!isModalOpen);\n  };\n\n  return (\n    <React.Fragment>\n      <Button variant=\"primary\" onClick={toggleModal}>\n        Show about modal\n      </Button>\n      <AboutModal\n        isOpen={isModalOpen}\n        onClose={toggleModal}\n        trademark=\"Trademark and copyright information here\"\n        brandImageSrc={brandImg}\n        brandImageAlt=\"Patternfly Logo\"\n        backgroundImageSrc={bgImg}\n      >\n        <TextContent>\n          <TextList component=\"dl\">\n            <TextListItem component=\"dt\">CFME Version</TextListItem>\n            <TextListItem component=\"dd\">5.5.3.4.20102789036450</TextListItem>\n          </TextList>\n        </TextContent>\n      </AboutModal>\n    </React.Fragment>\n  );\n};\n","title":"Custom background image","lang":"ts"}}>
      
    </Example>
};

const Component = () => (
  <React.Fragment>
    <AutoLinkHeader {...{"id":"examples","size":"h2","className":"ws-title ws-h2"}}>
      {`Examples`}
    </AutoLinkHeader>
    {React.createElement(pageData.examples["Basic"])}
    {React.createElement(pageData.examples["Without product name"])}
    {React.createElement(pageData.examples["Complex user positioned content"])}
    {React.createElement(pageData.examples["Custom background image"])}
  </React.Fragment>
);
Component.displayName = 'ComponentsAboutModalReactDocs';
Component.pageData = pageData;

export default Component;
