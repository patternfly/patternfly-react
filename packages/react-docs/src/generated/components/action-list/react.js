import React from 'react';
import { AutoLinkHeader, Example, Link as PatternflyThemeLink } from 'theme-patternfly-org/components';
import TimesIcon from '@patternfly/react-icons/dist/js/icons/times-icon';
import CheckIcon from '@patternfly/react-icons/dist/js/icons/check-icon';
const pageData = {
  "id": "Action list",
  "section": "components",
  "source": "react",
  "slug": "/components/action-list/react",
  "sourceLink": "https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/ActionList/examples/ActionList.md",
  "propComponents": [
    {
      "name": "ActionList",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list",
          "defaultValue": "''"
        },
        {
          "name": "isIconList",
          "type": "boolean",
          "description": "Flag indicating the action list contains multiple icons and item padding should be removed"
        }
      ]
    },
    {
      "name": "ActionListGroup",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list group"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list group",
          "defaultValue": "''"
        }
      ]
    },
    {
      "name": "ActionListItem",
      "description": "",
      "props": [
        {
          "name": "children",
          "type": "React.ReactNode",
          "description": "Children of the action list item"
        },
        {
          "name": "className",
          "type": "string",
          "description": "Additional classes added to the action list item",
          "defaultValue": "''"
        }
      ]
    }
  ],
  "cssPrefix": [
    "pf-c-action-list"
  ],
  "examples": [
    "Action list single group",
    "Action list with icons",
    "Action list multiple groups",
    "Action list with cancel button"
  ]
};
pageData.liveContext = {
  TimesIcon,
  CheckIcon
};
pageData.examples = {
  'Action list single group': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport {\n  ActionList,\n  ActionListItem,\n  Button,\n  Dropdown,\n  DropdownItem,\n  DropdownSeparator,\n  KebabToggle\n} from '@patternfly/react-core';\n\nexport const ActionListSingleGroup: React.FunctionComponent = () => {\n  const [isOpen, setIsOpen] = React.useState(false);\n\n  const onToggle = (\n    isOpen: boolean,\n    event: MouseEvent | TouchEvent | KeyboardEvent | React.KeyboardEvent<any> | React.MouseEvent<HTMLButtonElement>\n  ) => {\n    event.stopPropagation();\n    setIsOpen(isOpen);\n  };\n\n  const onSelect = (event: React.SyntheticEvent<HTMLDivElement, Event>) => {\n    event.stopPropagation();\n    setIsOpen(!isOpen);\n  };\n\n  const dropdownItems = [\n    <DropdownItem key=\"link\">Link</DropdownItem>,\n    <DropdownItem key=\"action\" component=\"button\">\n      Action\n    </DropdownItem>,\n    <DropdownItem key=\"disabled link\" isDisabled>\n      Disabled Link\n    </DropdownItem>,\n    <DropdownItem key=\"disabled action\" isDisabled component=\"button\">\n      Disabled Action\n    </DropdownItem>,\n    <DropdownSeparator key=\"separator\" />,\n    <DropdownItem key=\"separated link\">Separated Link</DropdownItem>,\n    <DropdownItem key=\"separated action\" component=\"button\">\n      Separated Action\n    </DropdownItem>\n  ];\n\n  return (\n    <React.Fragment>\n      <ActionList>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button\">\n            Back\n          </Button>\n        </ActionListItem>\n      </ActionList>\n      <br />\n      With kebab\n      <ActionList>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button2\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button2\">\n            Back\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Dropdown\n            onSelect={onSelect}\n            toggle={<KebabToggle onToggle={onToggle} />}\n            isOpen={isOpen}\n            isPlain\n            dropdownItems={dropdownItems}\n            position=\"right\"\n          />\n        </ActionListItem>\n      </ActionList>\n    </React.Fragment>\n  );\n};\n","title":"Action list single group","lang":"ts"}}>
      
    </Example>,
  'Action list with icons': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListItem, Button } from '@patternfly/react-core';\nimport TimesIcon from '@patternfly/react-icons/dist/esm/icons/times-icon';\nimport CheckIcon from '@patternfly/react-icons/dist/esm/icons/check-icon';\n\nexport const ActionListWithIcons: React.FunctionComponent = () => (\n  <ActionList isIconList>\n    <ActionListItem>\n      <Button variant=\"plain\" id=\"times-button\" aria-label=\"times icon button\">\n        <TimesIcon />\n      </Button>\n    </ActionListItem>\n    <ActionListItem>\n      <Button variant=\"plain\" id=\"check-button\" aria-label=\"check icon button\">\n        <CheckIcon />\n      </Button>\n    </ActionListItem>\n  </ActionList>\n);\n","title":"Action list with icons","lang":"ts"}}>
      
    </Example>,
  'Action list multiple groups': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListGroup, ActionListItem, Button } from '@patternfly/react-core';\n\nexport const ActionListMultipleGroups: React.FunctionComponent = () => (\n  <ActionList>\n    <ActionListGroup>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"next-button\">\n          Next\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"secondary\" id=\"back-button\">\n          Back\n        </Button>\n      </ActionListItem>\n    </ActionListGroup>\n    <ActionListGroup>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"submit-button\">\n          Submit\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"link\" id=\"cancel-button\">\n          Cancel\n        </Button>\n      </ActionListItem>\n    </ActionListGroup>\n  </ActionList>\n);\n","title":"Action list multiple groups","lang":"ts"}}>
      
    </Example>,
  'Action list with cancel button': props => 
    <Example {...pageData} {...props} {...{"code":"import React from 'react';\nimport { ActionList, ActionListGroup, ActionListItem, Button } from '@patternfly/react-core';\n\nexport const ActionListWithCancelButton: React.FunctionComponent = () => (\n  <React.Fragment>\n    In modals, forms, data lists\n    <ActionList>\n      <ActionListItem>\n        <Button variant=\"primary\" id=\"save-button\">\n          Save\n        </Button>\n      </ActionListItem>\n      <ActionListItem>\n        <Button variant=\"link\" id=\"cancel-button\">\n          Cancel\n        </Button>\n      </ActionListItem>\n    </ActionList>\n    <br />\n    In wizards\n    <ActionList>\n      <ActionListGroup>\n        <ActionListItem>\n          <Button variant=\"primary\" id=\"next-button\">\n            Next\n          </Button>\n        </ActionListItem>\n        <ActionListItem>\n          <Button variant=\"secondary\" id=\"back-button\">\n            Back\n          </Button>\n        </ActionListItem>\n      </ActionListGroup>\n      <ActionListGroup>\n        <ActionListItem>\n          <Button variant=\"link\" id=\"cancel-button2\">\n            Cancel\n          </Button>\n        </ActionListItem>\n      </ActionListGroup>\n    </ActionList>\n  </React.Fragment>\n);\n","title":"Action list with cancel button","lang":"ts"}}>
      
    </Example>
};

const Component = () => (
  <React.Fragment>
    <AutoLinkHeader {...{"id":"examples","size":"h2","className":"ws-title ws-h2"}}>
      {`Examples`}
    </AutoLinkHeader>
    {React.createElement(pageData.examples["Action list single group"])}
    {React.createElement(pageData.examples["Action list with icons"])}
    {React.createElement(pageData.examples["Action list multiple groups"])}
    {React.createElement(pageData.examples["Action list with cancel button"])}
  </React.Fragment>
);
Component.displayName = 'ComponentsActionListReactDocs';
Component.pageData = pageData;

export default Component;
