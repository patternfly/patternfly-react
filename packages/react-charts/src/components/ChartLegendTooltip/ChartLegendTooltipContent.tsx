import * as React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { Helpers, NumberOrCallback, StringOrNumberOrCallback } from 'victory-core';
import { VictoryLegend } from 'victory-legend';
import { ChartLabel } from '../ChartLabel';
import { ChartLegend } from '../ChartLegend';
import { ChartLegendTooltipLabel } from './ChartLegendTooltipLabel';
import { ChartLegendTooltipStyles, ChartThemeDefinition } from '../ChartTheme';
import {
  getLegendTooltipDataProps,
  getLegendTooltipSize,
  getLegendTooltipVisibleData,
  getLegendTooltipVisibleText,
  getTheme
} from '../ChartUtils';

/**
 * See https://github.com/FormidableLabs/victory/blob/master/packages/victory-core/src/index.d.ts
 * and https://github.com/FormidableLabs/victory/blob/master/packages/victory-legend/src/index.d.ts
 */
export interface ChartLegendTooltipContentProps {
  /**
   * The activePoints prop specifies the active data
   *
   * **This prop should not be set manually.**
   */
  activePoints?: any[];
  /**
   * The center prop determines the position of the center of the tooltip flyout. This prop should be given as an object
   * that describes the desired x and y svg coordinates of the center of the tooltip. This prop is useful for
   * positioning the flyout of a tooltip independent from the pointer. When ChartTooltip is used with
   * ChartVoronoiContainer, the center prop is what enables the mouseFollowTooltips option. When this prop is set,
   * non-zero pointerLength values will no longer be respected.
   */
  center?: { x: number; y: number };
  /**
   * Victory components can pass a datum prop to their label component. This can be used to calculate functional styles,
   * and determine child text
   */
  datum?: {};
  /**
   * The dx prop defines a horizontal shift from the x coordinate.
   */
  dx?: NumberOrCallback;
  /**
   * The dy prop defines a horizontal shift from the y coordinate.
   */
  dy?: NumberOrCallback;
  /**
   * The flyoutHeight prop defines the height of the tooltip flyout. This prop may be given as a positive number or a function
   * of datum. If this prop is not set, height will be determined based on an approximate text size calculated from the
   * text and style props provided to ChartTooltip.
   */
  flyoutHeight?: NumberOrCallback;
  /**
   * The flyoutWidth prop defines the width of the tooltip flyout. This prop may be given as a positive number or a
   * function of datum. If this prop is not set, flyoutWidth will be determined based on an approximate text size
   * calculated from the text and style props provided to VictoryTooltip.
   */
  flyoutWidth?: NumberOrCallback;
  /**
   * Specifies the height the svg viewBox of the chart container. This value should be given as a
   * number of pixels.
   *
   * Because Victory renders responsive containers, the width and height props do not determine the width and
   * height of the chart in number of pixels, but instead define an aspect ratio for the chart. The exact number of
   * pixels will depend on the size of the container the chart is rendered into.
   */
  height?: number;
  /**
   * The labelComponent prop takes a component instance which will be used
   * to render each legend label. The new element created from the passed
   * labelComponent will be supplied with the following properties: x, y,
   * style, and text. Any of these props may be overridden by passing in
   * props to the supplied component, or modified or ignored within the
   * custom component itself. If labelComponent is omitted, a new
   * ChartLabel will be created with the props described above.
   */
  labelComponent?: React.ReactElement<any>;
  /**
   * The legend component to render with chart.
   *
   * Note: Use legendData so the legend width can be calculated and positioned properly.
   * Default legend properties may be applied
   */
  legendComponent?: React.ReactElement<any>;
  /**
   * Specify data via the data prop. ChartLegend expects data as an array of objects with name (required), symbol, and
   * labels properties. The childName is used to sync the data series associated with the given chart child name.
   *
   * The data prop must be given as an array.
   *
   * @example legendData={[{ name: `GBps capacity - 45%` }, { name: 'Unused' }]}
   * @example legendData={[{ childName: `cats`, name: `Total cats` }, { childName: `dogs`, name: 'Total dogs' }]}
   */
  legendData?: {
    childName?: string;
    name?: string;
    labels?: {
      fill?: string;
    };
    symbol?: {
      fill?: string;
      type?: string;
    };
  }[];
  /**
   * The text prop defines the text ChartTooltip will render. The text prop may be given as a string, number, or
   * function of datum. When ChartLabel is used as the labelComponent, strings may include newline characters, which
   * ChartLabel will split in to separate <tspan/> elements.
   */
  text?: StringOrNumberOrCallback | string[] | number[];
  /**
   * The theme prop takes a style object with nested data, labels, and parent objects.
   * You can create this object yourself, or you can use a theme provided by
   * When using ChartLegend as a solo component, implement the theme directly on
   * ChartLegend. If you are wrapping ChartLegend in ChartChart or
   * ChartGroup, please call the theme on the outermost wrapper component instead.
   */
  theme?: ChartThemeDefinition;
  /**
   * Specifies the theme color. Valid values are 'blue', 'green', 'multi', etc.
   *
   * Note: Not compatible with theme prop
   *
   * @example themeColor={ChartThemeColor.blue}
   */
  themeColor?: string;
  /**
   * Specifies the theme variant. Valid values are 'dark' or 'light'
   *
   * Note: Not compatible with theme prop
   *
   * @example themeVariant={ChartThemeVariant.light}
   */
  themeVariant?: string;
  /**
   * The title prop specifies a title to render with the legend.
   * This prop should be given as a string, or an array of strings for multi-line titles.
   *
   * Valid types are string, string[], or Function
   *
   * Example: title={(datum) => datum.x}
   */
  title?: string | string[] | Function;
  /**
   * The titleComponent prop takes a component instance which will be used to render
   * a title for the component. The new element created from the passed
   * labelComponent will be supplied with the following properties: x, y, index, data,
   * datum, verticalAnchor, textAnchor, style, text, and events. Any of these props
   * may be overridden by passing in props to the supplied component, or modified
   * or ignored within the custom component itself. If labelComponent is omitted,
   * a new ChartLabel will be created with the props described above.
   */
  titleComponent?: React.ReactElement<any>;
  /**
   * Specifies the width of the svg viewBox of the chart container. This value should be given as a
   * number of pixels.
   *
   * Because Victory renders responsive containers, the width and height props do not determine the width and
   * height of the chart in number of pixels, but instead define an aspect ratio for the chart. The exact number of
   * pixels will depend on the size of the container the chart is rendered into.
   */
  width?: number;
}

export const ChartLegendTooltipContent: React.FunctionComponent<ChartLegendTooltipContentProps> = ({
  activePoints,
  center,
  datum,
  dx = 0,
  dy = 0,
  flyoutHeight,
  flyoutWidth,
  height,
  labelComponent = <ChartLegendTooltipLabel />,
  legendComponent = <ChartLegend />,
  legendData,
  text,
  themeColor,
  themeVariant,
  title,
  titleComponent = <ChartLabel />,
  width,

  // destructure last
  theme = getTheme(themeColor, themeVariant),
  ...rest
}: ChartLegendTooltipContentProps) => {
  const pointerLength = theme && theme.tooltip ? theme.tooltip.pointerLength : 10;
  const legendProps = getLegendTooltipDataProps(legendComponent.props);
  const visibleLegendData = getLegendTooltipVisibleData({
    activePoints,
    colorScale: legendProps.colorScale,
    legendData,
    text,
    theme
  });

  // Component offsets
  const legendOffsetX = 0;
  const legendOffsetY = title ? 5 : -10;
  const titleOffsetX = 10;
  const titleOffsetY = 0;

  // Returns x position of flyout
  const getX = () => {
    if (!(center || flyoutWidth || width)) {
      const x = (rest as any).x;
      return x ? x : undefined;
    }
    const _flyoutWidth = Helpers.evaluateProp(flyoutWidth);
    if (width > center.x + _flyoutWidth + pointerLength) {
      return center.x + ChartLegendTooltipStyles.flyout.padding / 2;
    } else if (center.x < _flyoutWidth + pointerLength) {
      return ChartLegendTooltipStyles.flyout.padding / 2 - pointerLength;
    } else {
      return center.x - _flyoutWidth;
    }
  };

  // Returns y position
  const getY = () => {
    if (!(center || flyoutHeight || height)) {
      const y = (rest as any).y;
      return y ? y : undefined;
    }
    const _flyoutHeight = Helpers.evaluateProp(flyoutHeight);
    if (center.y < _flyoutHeight / 2) {
      return ChartLegendTooltipStyles.flyout.padding / 2;
    } else if (center.y > height - _flyoutHeight / 2) {
      return height - _flyoutHeight + ChartLegendTooltipStyles.flyout.padding / 2;
    } else {
      return center.y - _flyoutHeight / 2 + ChartLegendTooltipStyles.flyout.padding / 2;
    }
  };

  // Min & max dimensions do not include flyout padding
  const maxLegendDimensions = getLegendTooltipSize({
    legendData: visibleLegendData,
    legendProps,
    text: getLegendTooltipVisibleText({ activePoints, legendData, text }),
    theme
  });
  const minLegendDimensions = getLegendTooltipSize({
    legendData: [{ name: '' }],
    legendProps,
    theme
  });

  // Returns the label component
  const getLabelComponent = () =>
    React.cloneElement(labelComponent, {
      dx: maxLegendDimensions.width - minLegendDimensions.width,
      legendData: visibleLegendData,
      ...labelComponent.props
    });

  // Returns the title component
  const getTitleComponent = () => {
    const _title = title instanceof Function ? title(datum) : title;

    return React.cloneElement(titleComponent, {
      style: {
        fill: ChartLegendTooltipStyles.label.fill,
        fontWeight: ChartLegendTooltipStyles.label.fontWeight
      },
      text: _title,
      textAnchor: 'start',
      x: getX() + titleOffsetX + Helpers.evaluateProp(dx),
      y: getY() + titleOffsetY + Helpers.evaluateProp(dy),
      ...titleComponent.props
    });
  };

  // Returns the legebd component
  const getLegendComponent = () =>
    React.cloneElement(legendComponent, {
      data: getLegendTooltipVisibleData({
        activePoints,
        colorScale: legendProps.colorScale,
        legendData,
        text,
        textAsLegendData: true,
        theme
      }),
      labelComponent: getLabelComponent(),
      standalone: false,
      theme,
      x: getX() + legendOffsetX + Helpers.evaluateProp(dx),
      y: getY() + legendOffsetY + Helpers.evaluateProp(dy),
      ...legendProps
    });

  return (
    <React.Fragment>
      {getTitleComponent()}
      {getLegendComponent()}
    </React.Fragment>
  );
};
ChartLegendTooltipContent.displayName = 'ChartLegendTooltipContent';

// Note: VictoryLegend.role must be hoisted, but getBaseProps causes error with ChartVoronoiContainer
hoistNonReactStatics(ChartLegendTooltipContent, VictoryLegend, { getBaseProps: true });
