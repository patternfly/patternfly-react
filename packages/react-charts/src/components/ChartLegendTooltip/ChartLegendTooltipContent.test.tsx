import * as React from 'react';
import { render } from '@testing-library/react';
import { ChartLegendTooltipContent } from './ChartLegendTooltipContent';

Object.values([true, false]).forEach(() => {
  test('ChartLegendTooltipContent', () => {
    const view = render(<ChartLegendTooltipContent />);
    expect(view.container).toMatchSnapshot();
  });
});

test('renders component text', () => {
  const legendData = [
    { name: 'Cats' },
    { name: 'Dogs', symbol: { type: 'dash' } },
    { name: 'Birds' },
    { name: 'Mice' }
  ];
  const view = render(<ChartLegendTooltipContent data={legendData} text={['1, 2, 3, 4']} />);
  expect(view.container).toMatchSnapshot();
});
