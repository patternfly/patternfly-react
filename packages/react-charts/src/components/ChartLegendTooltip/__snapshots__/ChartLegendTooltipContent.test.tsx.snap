// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartLegendTooltipContent 1`] = `
<div>
  <g>
    <rect
      height="0"
      role="presentation"
      shape-rendering="auto"
      style="fill: none;"
      vector-effect="non-scaling-stroke"
      width="0"
      x="NaN"
      y="NaN"
    />
  </g>
</div>
`;

exports[`ChartLegendTooltipContent 2`] = `
<div>
  <g>
    <rect
      height="0"
      role="presentation"
      shape-rendering="auto"
      style="fill: none;"
      vector-effect="non-scaling-stroke"
      width="0"
      x="NaN"
      y="NaN"
    />
  </g>
</div>
`;

exports[`renders component text 1`] = `
<div>
  <g>
    <rect
      height="0"
      role="presentation"
      shape-rendering="auto"
      style="fill: none;"
      vector-effect="non-scaling-stroke"
      width="0"
      x="NaN"
      y="NaN"
    />
  </g>
</div>
`;
